### PR TITLE
Feature/token bridge checkout

### PR DESCRIPTION
**Pull request by liayoo**
- Add checkout service for token bridging
  - _openCheckout
    - trigger path: `/checkout/request/${ainAddress}/${checkoutId}`
    - example value:
      ```
        {
           amount: 100,
           type: 'ETH', 
           token_id: '0xERC20_ADDR', 
           recipient: '0xETH_ADDR' 
        }
      ```
  - _closeCheckout
    - trigger path: `/checkout/history/${ainAddress}/${checkoutId}`
    - example value:
      ```
        {
           request: { ... }, // exactly the same as the corresponding open checkout request
           respones: {
             tx_hash: '0xETH_TX_HASH',
             status: 'SUCCESS',
           }
        }
      ```
    - note that writing history erases the corresponding request

- Introduce token bridge config
  - `/token/bridge/ETH/0x3A810ff7211b40c4fA76205a14efe161615d0385`:
    ```
    {
        "network_id": "1",
        "token_pool": "0x744Cb74A78Ac6dae46ebdaCa43e38ED60F965B8D",
        "min_checkout_per_request": 1000,
        "max_checkout_per_request": 10000,
        "max_checkout_per_day": 1000000,
        "token_exchange_rate": 1,
        "token_exchange_scheme": "FIXED"
    }
    ```

### CODE REVIEW

File: common/constants.js [hunk 1]
```diff
@@ -242,6 +244,13 @@ const PredefinedDbPaths = {
   CHECKIN_PAYLOAD: 'payload',
   CHECKIN_PARENT_FINALIZE: 'parent_finalize',
   CHECKOUT: 'checkout',
+  CHECKOUT_HISTORY: 'history',
+  CHECKOUT_HISTORY_REFUND: 'refund',
+  CHECKOUT_REQUEST: 'request',
```
Comment (platfowner) [inline common/constants.js:249]: What about plural form 'requests' as this path can contain multiple requests?

File: db/rule-util.js [hunk 2]
```diff
@@ -223,6 +223,48 @@ class RuleUtil {
...
+  validateCheckoutHistoryData(userAddr, checkoutId, data, getValue) {
+    const { FunctionResultCode } = require('../common/constants');
+    const request = getValue(`/checkout/request/${userAddr}/${checkoutId}`);
+    if (!request || !this.isDict(request) || !this.isDict(data)) {
+      return false;
+    }
+    if (!_.isEqual(request, data.request, { strict: true })) {
+      return false;
+    }
+    return this.isDict(data.response) && this.isString(data.response.tx_hash) &&
```
Comment (platfowner) [inline db/rule-util.js:264]: suggestion: what about adding a util function to rule-util and common-util for tx hash validation?
I guess stricter condition can be applied (e.g. length, hex string regex)?

File: db/functions.js [hunk 3]
```diff
@@ -981,6 +985,157 @@ class Functions {
...
+    if (incPendingResultCode !== true) {
+      return this.returnFuncResult(context, incPendingResultCode);
+    }
+    const {
+      token_pool: tokenPool,
+      min_checkout_per_request: minCheckoutPerRequest,
+      max_checkout_per_request: maxCheckoutPerRequest,
+      max_checkout_per_day: maxCheckoutPerDay,
+      token_exchange_rate: tokenExchangeRate,
+      token_exchange_scheme: tokenExchangeScheme,
```
Comment (platfowner) [inline db/functions.js:1073]: Could you define above param names in constants.js as e.g. TokenBridgeProperties?

File: db/functions.js [hunk 4]
```diff
@@ -981,6 +985,157 @@ class Functions {
...
+    }
+    return true;
+  }
+
+  validateTokenBridgeConfig(tokenPool, minCheckoutPerRequest, maxCheckoutPerRequest,
+      maxCheckoutPerDay, tokenExchangeRate, tokenExchangeScheme) {
+    if (tokenPool === undefined || minCheckoutPerRequest === undefined ||
+        maxCheckoutPerRequest === undefined || maxCheckoutPerDay === undefined ||
+        tokenExchangeRate === undefined || tokenExchangeScheme === undefined) {
+      return FunctionResultCode.FAILURE;
```
Comment (platfowner) [inline db/functions.js:1029]: Could you consider introducing more specific code e.g. INVALID_TOKEN_BRIDGE_CONFIG?

File: db/functions.js [hunk 5]
```diff
@@ -981,6 +985,157 @@ class Functions {
...
+    return true;
+  }
+
+  validateRecipient(recipient, tokenType) {
+    switch (tokenType) {
+      case 'AIN':
+      case 'ETH':
+        return CommonUtil.isCksumAddr(recipient) ? true : FunctionResultCode.FAILURE;
+    }
+    return FunctionResultCode.FAILURE;
```
Comment (platfowner) [inline db/functions.js:1040]: ditto e.g. INVALID_RECIPIENT?

File: db/functions.js [hunk 6]
```diff
@@ -981,6 +985,157 @@ class Functions {
...
+  }
+
+  validateCheckoutAmount(amount, timestamp, minCheckoutPerRequest, maxCheckoutPerRequest, maxCheckoutPerDay) {
+    const pendingTotal = this.db.getValue(PathUtil.getCheckoutPendingAmountTotalPath()) || 0; // includes 'amount'
+    const checkoutCompleteToday = this.db.getValue(
+        PathUtil.getCheckoutCompleteAmountDailyPath(CommonUtil.getDayTimestamp(timestamp))) || 0;
+    if (amount < minCheckoutPerRequest || amount > maxCheckoutPerRequest) {
+      return FunctionResultCode.FAILURE;
```
Comment (platfowner) [inline db/functions.js:1048]: ditto e.g. INVALID_CHECKOUT_AMOUNT?

File: db/functions.js [hunk 7]
```diff
@@ -981,6 +985,157 @@ class Functions {
...
+        tokenExchangeRate === undefined || tokenExchangeScheme === undefined) {
+      return FunctionResultCode.FAILURE;
+    }
+    return true;
+  }
+
+  validateRecipient(recipient, tokenType) {
+    switch (tokenType) {
+      case 'AIN':
+      case 'ETH':
```
Comment (platfowner) [inline db/functions.js:1037]: Are you planning to support both of them at once?
If not, why don't you leave the other as a todo?

File: db/functions.js [hunk 8]
```diff
@@ -981,6 +985,157 @@ class Functions {
...
+        return this.returnFuncResult(context, FunctionResultCode.FAILURE);
+      }
+    } else if (response.status === FunctionResultCode.SUCCESS) {
+      // Increase complete amounts
+      const updateStatsResultCode = this.updateCompleteCheckout(request.amount, context.timestamp, context);
+      if (updateStatsResultCode !== true) {
+        return this.returnFuncResult(context, updateStatsResultCode);
+      }
+    }
+    // Remove the original request
```
Comment (platfowner) [inline db/functions.js:1126]: Could you add as a NOTE more context on why this removal helps managing the requests?

File: integration/node.test.js [hunk 9]
```diff
@@ -6623,6 +6623,636 @@ describe('Blockchain Node', () => {
...
+        const totalPendingAmount = parseOrLog(syncRequest('GET',
+            server2 + `/get_value?ref=/checkout/stats/pending/total`)
+            .body.toString('utf-8')).result;
+        expect(afterRequestUserBalance).to.equal(beforeBalance - checkoutAmount);
+        expect(afterRequestTokenPoolBalance).to.equal(beforeTokenPoolBalance + checkoutAmount);
+        expect(userPendingAmount).to.equal(checkoutAmount);
+        expect(totalPendingAmount).to.equal(checkoutAmount);
+      });
+
+      it('non-authorized address cannot close checkout', async () => {
```
Comment (platfowner) [inline integration/node.test.js:6936]: minor: what about keeping 'cannot close checkout with <something>' pattern?

File: integration/node.test.js [hunk 10]
```diff
@@ -6623,6 +6623,636 @@ describe('Blockchain Node', () => {
...
+        expect(body.code).to.equal(1);
+        if (!(await waitUntilTxFinalized(serverList, _.get(body, 'result.tx_hash')))) {
+          console.error(`Failed to check finalization of tx.`);
+        }
+        const checkoutHistory = parseOrLog(syncRequest('GET',
+            server2 + `/get_value?ref=/checkout/history/${serviceUser}/0`).body.toString('utf-8')).result;
+        expect(checkoutHistory).to.equal(null);
+      });
+
+      it('token pool can close a successful checkout', async () => {
```
Comment (platfowner) [inline integration/node.test.js:6960]: minor: ditto e.g. 'can close checkout with token pool key'?

File: integration/node.test.js [hunk 11]
```diff
@@ -6623,6 +6623,636 @@ describe('Blockchain Node', () => {
...
+        const totalCompleteAmount = parseOrLog(syncRequest('GET',
+            server2 + `/get_value?ref=/checkout/stats/complete/total`)
+            .body.toString('utf-8')).result;
+        expect(userPendingAmount).to.equal(0);
+        expect(totalPendingAmount).to.equal(0);
+        expect(todayCompleteAmount).to.equal(checkoutAmount);
+        expect(totalCompleteAmount).to.equal(checkoutAmount);
+      });
+
+      it('token pool can close a failed checkout and refund', async () => {
```
Comment (platfowner) [inline integration/node.test.js:7069]: minor: ditto e.g. 'can close and refund failed checkout'?

File: common/constants.js (hunk 1 above)
Comment (liayoo) [reply common/constants.js]: Done!

File: db/rule-util.js (hunk 2 above)
Comment (liayoo) [reply db/rule-util.js]: Added `isValidHash()`. Thanks for the idea! 👍

File: db/functions.js (hunk 3 above)
Comment (liayoo) [reply db/functions.js]: Done!

File: db/functions.js (hunk 4 above)
Comment (liayoo) [reply db/functions.js]: Done!

File: db/functions.js (hunk 5 above)
Comment (liayoo) [reply db/functions.js]: Done

File: db/functions.js (hunk 6 above)
Comment (liayoo) [reply db/functions.js]: Done

File: db/functions.js (hunk 7 above)
Comment (liayoo) [reply db/functions.js:1036]: I was going to add 'AIN' type for the shards, but I'll leave it as a todo for now and implement it when we revamp sharding.

File: db/functions.js (hunk 8 above)
Comment (liayoo) [reply db/functions.js]: Right, added a more detailed explanation :)

File: integration/node.test.js (hunk 9 above)
Comment (liayoo) [reply integration/node.test.js]: Done!

File: integration/node.test.js (hunk 10 above)
Comment (liayoo) [reply integration/node.test.js]: Done

File: integration/node.test.js (hunk 11 above)
Comment (liayoo) [reply integration/node.test.js]: Done!

**Review by platfowner — APPROVED** [review]
LGTM
Thanks for the nice implementation.

**Comment by liayoo** [comment]
Thanks for the review! Now merging..

**Pull request merged**